[JSC] Implement `Set.prototype.intersection` in C++

sosukesuzuki opened the pull request:
#### 736aa4a741da8973040b68b30bb002d3752ef89b
<pre>
[JSC] Implement `Set.prototype.intersection` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=295476">https://bugs.webkit.org/show_bug.cgi?id=295476</a>

Reviewed by Yusuke Suzuki.

This patch implements `Set.prototype.intersection` in C++. The C++
implementation enters a fast path when `otherSet` is in an ideal state.
This significantly improves performance, especially in the baseline JIT.
There are also some performance improvements in DFG/FTL.

Baseline:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this      2794.4450+-41.3768    ^    936.5865+-34.5267       ^ definitely 2.9836x faster
set-prototype-intersection-large-other     1930.2733+-23.6332    ^    927.6492+-26.7125       ^ definitely 2.0808x faster
set-prototype-intersection                 2937.9777+-19.6155    ^   1152.9432+-36.6889       ^ definitely 2.5482x faster

with DFG only:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this      1386.4293+-24.1653    ^    915.1607+-10.0156       ^ definitely 1.5150x faster
set-prototype-intersection-large-other     1263.0578+-9.4375     ^    932.4156+-30.8533       ^ definitely 1.3546x faster
set-prototype-intersection                 1794.7610+-24.5289    ^   1143.6836+-25.1361       ^ definitely 1.5693x faster

with DFG and FTL:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this       994.4368+-28.4201    ^    927.9790+-17.3331       ^ definitely 1.0716x faster
set-prototype-intersection-large-other     1117.0381+-12.4878    ^    930.5520+-19.9727       ^ definitely 1.2004x faster
set-prototype-intersection                 1553.9095+-16.2586    ^   1147.2663+-38.2916       ^ definitely 1.3544x faster

* JSTests/microbenchmarks/set-prototype-intersection-large-other.js: Added.
(shouldBe):
* JSTests/microbenchmarks/set-prototype-intersection-large-this.js: Added.
(shouldBe):
* JSTests/microbenchmarks/set-prototype-intersection.js: Added.
(shouldBe):
* JSTests/stress/set-prototype-intersection-larger-other-effects-1.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-other-effects-2.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-this-effects-1.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-this-effects-2.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::setPrimordialPropertiesWatchpointSet):
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):
(JSC::getSetSizeAsInt):
(JSC::fastSetIntersection):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetPrototypeInlines.h:
(JSC::setPrimordialWatchpointIsValid):

Canonical link: <a href="https://commits.webkit.org/297882@main">https://commits.webkit.org/297882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f255ffc055305502736eb3ec708df8e0e6b644ca

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113166 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32901 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23379 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119374 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/63938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33553 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41464 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86140 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41146 "Unexpected infrastructure issue, retrying build") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101803 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66460 "Found 142 new API test failures: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestBackForwardList:/webkit/WebKitWebView/navigation-after-session-restore, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/reset, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WPE/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone-worker, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/injected-script, /WPE/TestLoaderClient:/webkit/WebKitWebView/unfinished-subresource-load, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cookies ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19931 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63128 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105683 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96186 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122591 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111782 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30024 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94991 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94733 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24183 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39876 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17682 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36376 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40130 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45629 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136012 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39771 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36498 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41508 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -175,6 +181,170 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+                ASSERT(!args.hasOverflowed());
+                hasResult = call(globalObject, has, hasCallData, otherValue, args);
+                RETURN_IF_EXCEPTION(scope, { });
+            }
```
`CachedCall` is faster when we set up this outside of the loop.

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -175,6 +181,175 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSSet* thisSet = getSet(globalObject, callFrame->thisValue());
+    RETURN_IF_EXCEPTION(scope, { });
```
This is not necessary.

sosukesuzuki replied:
If the second argument is not a Set object, `getSet` throws a `TypeError` and returns `nullptr`, so I think this is necessary. In fact, if `RETURN_IF_EXCEPTION` is removed, a segmentation fault will occur.

darinadler replied:
Just curious: Which of the tests in this pull request covers that case?

sosukesuzuki replied:
test262 includes that case ( Of course, I can add such cases to stress tests )

Constellation replied:
Oh, that's right.

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -175,6 +181,175 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+            scope.release();
+            return fastSetIntersection(globalObject, thisSet, otherSet);
+        }
+    }
```
I think we should first check `auto* otherSet = jsDynamicCast<JSSet*>(otherValue.asCell())`

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -175,6 +181,175 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+            CachedCall cachedCall(globalObject, jsCast<JSFunction*>(keys), 0);
+            RETURN_IF_EXCEPTION(scope, { });
+            iterator = cachedCall.callWithArguments(globalObject, otherValue);
+            RETURN_IF_EXCEPTION(scope, { });
```
If we do not call it multiple times, we should just use `call` (not CachedCall).

sosukesuzuki committed:
[JSC] Implement `Set.prototype.intersection` in C++
https://bugs.webkit.org/show_bug.cgi?id=295476

Reviewed by Yusuke Suzuki.

This patch implements `Set.prototype.intersection` in C++. The C++
implementation enters a fast path when `otherSet` is in an ideal state.
This significantly improves performance, especially in the baseline JIT.
There are also some performance improvements in DFG/FTL.

Baseline:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this      2794.4450+-41.3768    ^    936.5865+-34.5267       ^ definitely 2.9836x faster
set-prototype-intersection-large-other     1930.2733+-23.6332    ^    927.6492+-26.7125       ^ definitely 2.0808x faster
set-prototype-intersection                 2937.9777+-19.6155    ^   1152.9432+-36.6889       ^ definitely 2.5482x faster

with DFG only:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this      1386.4293+-24.1653    ^    915.1607+-10.0156       ^ definitely 1.5150x faster
set-prototype-intersection-large-other     1263.0578+-9.4375     ^    932.4156+-30.8533       ^ definitely 1.3546x faster
set-prototype-intersection                 1794.7610+-24.5289    ^   1143.6836+-25.1361       ^ definitely 1.5693x faster

with DFG and FTL:
                                                TipOfTree                  Patched

set-prototype-intersection-large-this       994.4368+-28.4201    ^    927.9790+-17.3331       ^ definitely 1.0716x faster
set-prototype-intersection-large-other     1117.0381+-12.4878    ^    930.5520+-19.9727       ^ definitely 1.2004x faster
set-prototype-intersection                 1553.9095+-16.2586    ^   1147.2663+-38.2916       ^ definitely 1.3544x faster

* JSTests/microbenchmarks/set-prototype-intersection-large-other.js: Added.
(shouldBe):
* JSTests/microbenchmarks/set-prototype-intersection-large-this.js: Added.
(shouldBe):
* JSTests/microbenchmarks/set-prototype-intersection.js: Added.
(shouldBe):
* JSTests/stress/set-prototype-intersection-larger-other-effects-1.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-other-effects-2.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-this-effects-1.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* JSTests/stress/set-prototype-intersection-larger-this-effects-2.js: Added.
(shouldBe):
(get hasGet):
(get keysGet):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::setPrimordialPropertiesWatchpointSet):
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):
(JSC::getSetSizeAsInt):
(JSC::fastSetIntersection):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetPrototypeInlines.h:
(JSC::setPrimordialWatchpointIsValid):

Canonical link: https://commits.webkit.org/297882@main
